### PR TITLE
feat(fulfillment): Separate list and context rules validation

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

- Add method to validate fulfillment option from the provider
- Separate list/list and count from context rules validation and add listShippingOptionsForContext

FIXES CORE-1861

### CODE REVIEW

File: packages/fulfillment/src/services/fulfillment-module-service.ts
Comment (olivermrbl) [inline packages/fulfillment/src/services/fulfillment-module-service.ts]: question: Would it make sense to call it `listShippingOptionsForContext` and have the same behavior that `listShippingOptions` had before?

Comment (olivermrbl) [reply packages/fulfillment/src/services/fulfillment-module-service.ts]: Or `listValidShippingOptions`

Comment (adrien2p) [reply packages/fulfillment/src/services/fulfillment-module-service.ts]: I prefer the first one, i can do that

File: packages/fulfillment/src/services/fulfillment-module-service.ts
Comment (sradevski) [inline packages/fulfillment/src/services/fulfillment-module-service.ts]: Why do we need the custom implementation here? I know there is a good reason, but what is it? 😄

**Review by sradevski — APPROVED** [review]

File: packages/fulfillment/src/services/fulfillment-module-service.ts
Comment (olivermrbl) [reply packages/fulfillment/src/services/fulfillment-module-service.ts]: With the separate method `listShippingOptionsForContext` maybe we don't need this?

I would expect `listShippingOptions` to handle filters the same way other list methods do.

Comment (adrien2p) [reply packages/fulfillment/src/services/fulfillment-module-service.ts]: Ahah there is no good reason, it is just because of the shortcut fulfillment_set_id and so on which means that we need to re map them.
But i ll wait until we settle on the approach and then cleanup.

File: packages/fulfillment/src/services/fulfillment-module-service.ts
Comment (adrien2p) [reply packages/fulfillment/src/services/fulfillment-module-service.ts]: I am wondering if it makes sense to do that in truth. Why two methods one to list and one to list plus filter on context when we had one method to do this and if you dont pass the context it just performs a classic list. What do you prefer? But i think that if we have a separate method for rules validation then it shouldn't list otherwise it comes back to what we had before no?

**Comment by olivermrbl** [comment]
I might have caused some confusion. After giving this some thought, here's what I think we should do.

I would suggest we have the following methods:

`listShippingOptions`

`listAndCountShippingOptions`

`validateShippingOption`

None of the methods should expose a different filter object from what we usually do i.e. I think we should remove the normalization.

The first two methods would behave like any other `list*` method, with the exception of the context parameter. If the context is passed, it performs an additional validation to ensure only the shipping options applicable for the given context are returned. As you suggest @adrien2p. 

The latter would be a convenient method, wrapping the `listShippingOptions`. 

```ts
async validateShippingOption(option, context) {
  const [shippingOption] = await this.listShippingOptions({ id: option.id }, context)

  if (shippingOptions) {
    return true
  }

  return false
}
```

Whether or not we add the third method I don't have a strong opinion about. Just think it's nice to call a dedicated method for the purpose of validating the applicableness of options.

For example, I will need to add a `refreshShipping` step in all cart workflows to ensure the currently created shipping methods are still applicable after the update. If not, they should be deleted.

**Comment by adrien2p** [comment]
@olivermrbl I ve done the changes as per the discussion, let me know what you think

File: packages/fulfillment/src/services/fulfillment-provider.ts
Comment (sradevski) [inline packages/fulfillment/src/services/fulfillment-provider.ts:81]: Do we even need this in the custom service, or should we just put this logic in `fulfillment-module-service` instead?

**Review by sradevski — APPROVED** [review]

File: packages/fulfillment/src/services/fulfillment-provider.ts
Comment (adrien2p) [reply packages/fulfillment/src/services/fulfillment-provider.ts:81]: This is a service that communicate with the third party provider installed on the module, everything link to third party should be separated from the module service itself IMO, wdyt?

**Comment by srindom** [comment]
Sorry for chiming in late - I have one additional method that I'd like the module to have:

```
getShippingOptionsForContext(fulfillmentSetIds, fulfillmentContext)
```

This is a convenience method on top of `listShippingOptions` + it verifies that options are valid for the context. This will be the primary function used when figuring out shipping options applicable to a cart. 

Also want to encourage that `listShippingOptions` doesn't know the context but instead behaves exactly like all other `list*` methods throughout our modules. 

The motivation for this is:

1. It simplifies the work the module consumer has to do to get shipping options. (e.g., you don't need to remember the exact data structure of service_zones -> geo_zones).
2. It allows us to validate and throw if the context passed is insufficient for finding shipping options.
3. It provides a single API that we can innovate behind (e.g., we can add new features in the context or similar without requiring people to update their `list` params.
4. We keep behavior of all `list*` calls the same whic

**Comment by adrien2p** [comment]
@srindom yes it makes perfect sense. What do you think of the current changes? Also, I am waiting for oli to be available to discuss something around the failing test. then it should be good to be merged if you approve the changes

File: packages/fulfillment/src/services/fulfillment-module-service.ts
Comment (olivermrbl) [inline packages/fulfillment/src/services/fulfillment-module-service.ts:1501]: love it!

**Review by olivermrbl — APPROVED** [review]
great work @adrien2p , this lgtm

**Comment by adrien2p** [comment]
@srindom ill wait for your approval before putting the automerge label

**Pull request merged**
